### PR TITLE
Clone develop 2

### LINKED ISSUE #10: As a game user, I want to choose my play style and strategies based on the situation so that I can use strategies during gameplay
Core feature: Playstyle and strategy select option through the specialized class system - sprint 4

Story Points: 4
Risk: Medium

## Acceptance Test # 1 - Player Choice: 
1. Start the game
2. Create a room
               2.1 Choose one of the options of the class for the specific playstyle
                              2.1.1. Mage Class, or
                              2.1.2 Defender Class
3. Check the towers displayed
         
### During the game: 

The Player should wee the list of towers specific for their class:
- Mage class will have Base Tower, Fire Tower, and Light Tower
- Defender class will have Base Tower, Slow Tower, and Light Tower

## Acceptance Test # 2 - Tower Debuffs: 
1. Start the game
2. Place towers on the map prior to the beginning of the wave
               2.1 Choose the type of tower you want to be placed
                              2.1.1. Fire Tower: Does damage to the enemies
                              2.1.2 Slow Tower: Does less damage, but reduces the speed of the enemy
                              2.1.3 Light Tower: Does more damage to the enemies than all of the other towers
3. Start the wave
         
### During the wave: 

- The Towers will apply a specific amount of damage to the enemies and apply debuffs if applicable for the type
-  Different positionings of the towers of various types may make it easier to eliminate the enemies

### LINKED ISSUE #94: [TASK] Adapting upgrade tower with network
Details: Add network function for upgrading tower features.

Story Points: 4 (2 points to each member)
Risk: High

## Acceptance Test:
1. Build a client (client1) and start a client on Unity Editor (client2).
2. Start the game on both clients ( username, create and join room steps should be done without error). Let client1 create the room, and client2 join the room.
3. Let client1 place a tower(tower1), tower1 should be correctly displayed on both clients.
4. Let client1 select tower1. The Select Panel should appear on client1 (bottom right scene) and not appear on client2.
5. Let client1 hover the mouse on upgrade button (middle). The tooltip should appear with tower1's stats on bottom left of the scene. This tooltip should not appear on client2 scene.
6. Let client1 click upgrade button. tower1's stats should be changed and can be checked by Inspector on client2.
7. The currency display should correctly change by decreasing the same amount for upgrading tower1 on both clients.
8. The price text on sell button of tower1 should correctly change.
9. Let client1 sell tower1.
10. The currency display should correctly change by increasing the same amount for selling tower1 on both clients.
11. Do step1 to step 10 again but switch role of client1 and client2 and use different type of tower.

### LINKED ISSUE #106: [TASK] Fixing projectiles bug
Details: Fixing bug related to projectiles that are not synchronized over the network.
Bug detail:
- Projectiles, were created by the masterclient's towers, appeared in the middle of the scene on client side.
- Projectiles did not move to the monster on client side,
- This bug is involved with the way of 'Instantiate Objects' over the network which related to Towers, Projectiles, Monsters.  

Attempt:
- Change the way of instantiating for Towers, Projectiles, Monsters by initializing these objects with OnPhotonInstantiate() function from Photon.Pun package.

Story Points: 5
Risk: High

## Acceptance Test:
1. Start the game.
2. Place each kind of towers on the map prior to the beginning of the wave, the towers should appear at the same place on both clients.
3. Start the wave.
4. Monsters should appear at the same place on both clients.
5. Monsters should move at the same speed on both clients.
6. Monsters should reach a range of a tower at the same time on both clients.
7. Towers should shoot projectiles at the same time on both clients.
8. Projectiles should appear at the same time on both clients.
9. Projectiles should move at the same speed to the same target monster on both clients.
10. Projectiles should hit the target monster at the same time on both clients.
11. Monsters should take the same damage from projectiles on both clients.
12. Monsters' HP should reduce with the same amount on both clients.
12. Monsters should be destroyed (when HP is 0 or reach the base) at the same time on both clients.

### LINKED ISSUE #107: [TASK] Adapting leveling system with network
Details: Add network function for leveling system feature.

Story Points: 3
Risk: High

## Acceptance Test:
1. Build a client (client1) and start a client on Unity Editor (client2).
2. Start the game on both clients ( username, create and join room steps should be done without error). Let client1 create the room, and client2 join the room.
3. client1 will place a tower(tower1), client2 will place a tower(tower2). Both towers should be displayed correctly on both players' scene.
4. client1 click tower1. The Select Panel should appear on client1 (bottom right scene) and not appear on client2.
5. client1 hover to tech upgrade button. The tooltip panel should appear with the stats on bottom left of the scene.
6. client1.click tech upgrade button. Nothing should happen to the tower1 (its stats should not be changed).
7. Start the wave from client1.
8. Monsters should appear and be destroyed by towers. Exp display should be correctly changed whenever the monster is destroyed on both clients. (Scarecrow gives 50exp, TrainningDummy give 25exp)
9. If the exp is enough to level up. The Level display should be correctly gone up by 1 and Exp display should be reset back to 0 on both clients.
10. If the level is up by 1. The Tokens display should be gone up by 1 too on both clients.
11. client1 click tower1. The Select Panel should appear on client1 (bottom right scene). 
12. client1 click tech upgrade button (right most left button). Tower1's stats should be correctly changed. On client2, the tower1's stats should be changed too and can be checked from the Inspector in Editor.
13. Check if The Tokens display on client1 should be gone down by 1. The tokens display on client2 should not be changed.
14. Check if The tech upgrade button should be display as 'max'.
15. Do step 1 to 14 again but on step 2 client2 will create the room, and client1 will join the room.

### LINKED ISSUE #108: [Task] Adapting Tower's Debuff functions over network
Details: Add network functions for tower's debuff features.

Story Points: 1
Risk: High

## Acceptance Test:
1. Build a client (client1) and start a client from Unity Editor (client2).
2. Start the game on both clients ( username, create and join room steps should be done without error). Let client1 create the room, and client2 join the room.
3. Let client1 pick Mage Class and client2 pick Defender Class.
4. Let client1 place FireTower.
5. Let client1 start the wave.
6. Check if FireProjectile causes debuff to the monster on hit on client1. If yes, the monster's HP should be decreased overtime on both clients. If no, the monster's HP should not be changed on both clients.
8. Check if debuff on the monster wear off. If yes, the monster's HP should stop decreasing on both clients. If no, the monster's HP should keep decreasing overtime on both clients.
7. Let client1 sell FireTower.
8. Let client1 place LightTower.
9. Check if LightProjectile causes debuff to the monster on hit on client1. If yes, the monster should stop moving on both clients. If no, the monster should keep moving normally on both clients.
10. Check if debuff on the monster wear off. If yes, the monster should start moving again on both clients. If no, the monster should keep stopping on both clients.
11. Let client1 sell LightTower.
12. Let client2 place SlowTower.
13. Check if SlowProjectile causes debuff to the monster on hit on client2. If yes, the monster's speed should decreasing overtime on both clients. If no, the monster's speed should not be changed on both clients.
14. Check if debuff on the monster wear off. If yes, the monster's speed should stop decreasing on both clients. If no, the monster's speed should keep decreasing overtime on both clients.
15. Do step 1 to 14 again but switch role for client1 and client2.

### PR DESCRIPTION
**Pull request by Phamhhuy**
All conflicted files from Adapting-Upgrade-Tower-With-Coop branch have been solved.
Merged the Adapting-Upgrade-Tower-With-Coop branch to clone_develop_2 branch.
Now, PR to merge from clone_develop_2 branch to develop branch

### CODE REVIEW

File: LeagueOfTowers/Assets/Scripts/CurrencyManager.cs [hunk 1]
```diff
@@ -10,7 +10,7 @@
 public class CurrencyManager : Singleton<CurrencyManager>
 {
     //an event that is triggered when the currency changes
-    public event CurrencyChanged Changed;
+    //public event CurrencyChanged Changed;
```
Comment (nuridak) [inline LeagueOfTowers/Assets/Scripts/CurrencyManager.cs]: I may be wrong, but if these variables are not needed anymore, wouldn't it be better if we delete them?

File: LeagueOfTowers/Assets/Scripts/CurrencyManager.cs [hunk 2]
```diff
@@ -66,7 +66,7 @@ public void AddCurrency(int earnAmount)
         this.view.RPC("addCurrencyRPC", RpcTarget.All, earnAmount);
 
         //call currency change
-        OnCurrencyChanged();
+        //OnCurrencyChanged();
```
Comment (nuridak) [inline LeagueOfTowers/Assets/Scripts/CurrencyManager.cs]: Same with this, do we need this?

File: LeagueOfTowers/Assets/Scripts/FireTowerScript.cs [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+        base.Start();
+        //base.Set();
+        //setElementType(Element.FIRE);
+
+        // Upgrades = new TowerUpgrade[]
+        // {
+        //     //price, damage, debuffDuration, debuffProcChance, tickTime, damage per tick
+        //     new TowerUpgrade(10, 5, 0.5f, 5, 0.1f, 1),
+        //     new TowerUpgrade(15, 10, 0.5f, 5, 0.1f, 1),
+        // };
```
Comment (nuridak) [inline LeagueOfTowers/Assets/Scripts/FireTowerScript.cs]: Is this part needed?

File: LeagueOfTowers/Assets/Scripts/FireTowerScript.cs [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    // }
+
+    //override the default upgrade tower function
+    // public override void Upgrade()
+    // {
+    //     //decrease ticktime and increase tick damage
+    //     this.tickTime -= GetNextUpgrade.TickTime;
+    //     this.tickDamage += GetNextUpgrade.TickDamage;
+    //     base.Upgrade();
+    // }
```
Comment (nuridak) [inline LeagueOfTowers/Assets/Scripts/FireTowerScript.cs]: commented code

File: LeagueOfTowers/Assets/Scripts/LevelUpManager.cs [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    //actions
+    public void AddExpPoints(int earnEXP)
+    {
+        // this.expPoints += earnEXP;
+        // this.expPointsDisplayText.text = string.Format("Exp: {0} / {1}", this.expPoints.ToString(), this.expRequireToLevelUP);
+
+        // if(this.expPoints >= this.expRequireToLevelUP)
+        // {
+        //     LevelUp();
+
```
Comment (nuridak) [inline LeagueOfTowers/Assets/Scripts/LevelUpManager.cs]: Commented code

File: LeagueOfTowers/Assets/Scripts/LightTowerScript.cs [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        base.Start();
+        // base.Set();
+        // setElementType(Element.LIGHT);
+
+        // Upgrades = new TowerUpgrade[]
+        // {
+        //     //price, damage, debuffDuration, debuffProcChance
+        //     new TowerUpgrade(10, 5, 1f, 5),
+        //     new TowerUpgrade(15, 10, 1f, 5),
+        // };
```
Comment (nuridak) [inline LeagueOfTowers/Assets/Scripts/LightTowerScript.cs]: Commented code

File: LeagueOfTowers/Assets/Scripts/LightTowerScript.cs [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    }
+
+    //override and also add tower header string
+    // public override string GetTechStats()
+    // {
+    //     //return a string result with tower header
+    //     string result = string.Format("<color=#FFFF00>{0}</color> {1}", tooltipHeader, base.GetTechStats());
+
+    //     return result;
+    // }
```
Comment (nuridak) [inline LeagueOfTowers/Assets/Scripts/LightTowerScript.cs]: Commented code

File: LeagueOfTowers/Assets/Scripts/ProjectileScript.cs [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+                DestroyThisProjectile();
+            }
+            
+        }
+        else{
+            DestroyThisProjectile();
+        }
+        // else{
+        //     Debug.Log("not found");
+        // }
```
Comment (nuridak) [inline LeagueOfTowers/Assets/Scripts/ProjectileScript.cs]: Is there an option to leave things like this, but use them only in the Debug mode?
Similar things noted with "Debug"

File: LeagueOfTowers/Assets/Scripts/ProjectileScript.cs [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+
+    public int GetProjectileViewID()
+    {
+        return this.view.ViewID;
+    }
+
+    //events
+    public void OnPhotonInstantiate(PhotonMessageInfo info)
+    {
+        //Debug.Log("called!");
```
Comment (nuridak) [inline LeagueOfTowers/Assets/Scripts/ProjectileScript.cs:60]: Debug

File: LeagueOfTowers/Assets/Scripts/ProjectileScript.cs [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+        object[] data = this.gameObject.GetPhotonView().InstantiationData;
+        if(data != null && data.Length == 5){
+            //projectileType == projectileName
```
Comment (nuridak) [inline LeagueOfTowers/Assets/Scripts/ProjectileScript.cs:63]: commented code

File: LeagueOfTowers/Assets/Scripts/SlowTowerScript.cs [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+        // base.Set();
+        // setElementType(Element.SLOW);
+
+        // Upgrades = new TowerUpgrade[]
+        // {
+        //     //price, damage, debuffDuration, debuffProcChance, slowing factor
+        //     new TowerUpgrade(10, 5, 1f, 5, 10),
+        //     new TowerUpgrade(15, 10, 1f, 5, 20),
+        // };
+    }
```
Comment (nuridak) [inline LeagueOfTowers/Assets/Scripts/SlowTowerScript.cs]: commented code

File: LeagueOfTowers/Assets/Scripts/SlowTowerScript.cs [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+    //     return result;
+    // }
+
+    //override the default upgrade tower function
+    // public override void Upgrade()
+    // {
+    //     //increase the slowing factor
+    //     this.slowFactor += GetNextUpgrade.SlowingFactor;
+    //     base.Upgrade();
+    // }
```
Comment (nuridak) [inline LeagueOfTowers/Assets/Scripts/SlowTowerScript.cs]: commented code

File: LeagueOfTowers/Assets/Scripts/SpawnPlayerManager.cs [hunk 13]
```diff
@@ -2,23 +2,50 @@
...
 
     // Update is called once per frame
     private void Update()
     {
         
     }
+
+    public void MageClassPicked()
+    {
+        Debug.Log("Mage class created");
```
Comment (nuridak) [inline LeagueOfTowers/Assets/Scripts/SpawnPlayerManager.cs]: Debug

File: LeagueOfTowers/Assets/Scripts/TowerButton.cs [hunk 14]
```diff
@@ -12,14 +12,16 @@ public class TowerButton : MonoBehaviour
...
+
 
     // Start is called before the first frame update
     private void Start()
     {
         this.SetPriceText();
 
         //add a new currency changed event to the tower buttons
-        CurrencyManager.Instance.Changed += new CurrencyChanged(PriceCheck);
+        //CurrencyManager.Instance.Changed += new CurrencyChanged(PriceCheck);
```
Comment (nuridak) [inline LeagueOfTowers/Assets/Scripts/TowerButton.cs]: commented line

**Review by nuridak — COMMENTED** [review]
Everything worked on the computer :) 
There is a lot of commented code and Lines for Debug, Should we delete all of the commented code since we do not use it anymore to make it look cleaner?

File: LeagueOfTowers/Assets/Scripts/CurrencyManager.cs (hunk 1 above)
Comment (Phamhhuy) [reply LeagueOfTowers/Assets/Scripts/CurrencyManager.cs]: i removed.

File: LeagueOfTowers/Assets/Scripts/CurrencyManager.cs (hunk 2 above)
Comment (Phamhhuy) [reply LeagueOfTowers/Assets/Scripts/CurrencyManager.cs]: i removed.

File: LeagueOfTowers/Assets/Scripts/FireTowerScript.cs (hunk 3 above)
Comment (Phamhhuy) [reply LeagueOfTowers/Assets/Scripts/FireTowerScript.cs]: i removed.

File: LeagueOfTowers/Assets/Scripts/FireTowerScript.cs (hunk 4 above)
Comment (Phamhhuy) [reply LeagueOfTowers/Assets/Scripts/FireTowerScript.cs]: i removed.

File: LeagueOfTowers/Assets/Scripts/LevelUpManager.cs (hunk 5 above)
Comment (Phamhhuy) [reply LeagueOfTowers/Assets/Scripts/LevelUpManager.cs]: i removed.

File: LeagueOfTowers/Assets/Scripts/ProjectileScript.cs (hunk 10 above)
Comment (Phamhhuy) [reply LeagueOfTowers/Assets/Scripts/ProjectileScript.cs:63]: I need to keep that comment to remind me that this projectileType field is also a projectile's name.

File: LeagueOfTowers/Assets/Scripts/SlowTowerScript.cs (hunk 11 above)
Comment (Phamhhuy) [reply LeagueOfTowers/Assets/Scripts/SlowTowerScript.cs]: i removed.

File: LeagueOfTowers/Assets/Scripts/SlowTowerScript.cs (hunk 12 above)
Comment (Phamhhuy) [reply LeagueOfTowers/Assets/Scripts/SlowTowerScript.cs]: i removed.

File: LeagueOfTowers/Assets/Scripts/SpawnPlayerManager.cs (hunk 13 above)
Comment (Phamhhuy) [reply LeagueOfTowers/Assets/Scripts/SpawnPlayerManager.cs]: i put it in comment.

File: LeagueOfTowers/Assets/Scripts/TowerButton.cs (hunk 14 above)
Comment (Phamhhuy) [reply LeagueOfTowers/Assets/Scripts/TowerButton.cs]: i removed.

File: LeagueOfTowers/Assets/Scripts/ProjectileScript.cs (hunk 10 above)
Comment (nuridak) [reply LeagueOfTowers/Assets/Scripts/ProjectileScript.cs:63]: I see

**Review by nuridak — APPROVED** [review]

**Pull request merged**